qualtran/bloqs/state_preparation/state_preparation_alias_sampling.py

```diff
@@ -235,6 +235,214 @@ def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+            mu=mu,
+            sum_of_lcu_coeffs=sum_of_lcu_coeffs,
+        )
+
+    @property
+    def n_coeff(self) -> SymbolicInt:
+        return self.selection_registers[0].dtype.iteration_length_or_zero()
+
+    @property
+    def n_nonzero_coeff(self) -> SymbolicInt:
+        return slen(self.index)
+
+    @cached_property
+    def l1_norm_of_coeffs(self) -> 'SymbolicFloat':
+        return self.sum_of_lcu_coeffs
+
+    @cached_property
+    def sigma_mu_bitsize(self) -> SymbolicInt:
+        return self.mu
+
+    @cached_property
+    def sparse_index_bitsize(self) -> SymbolicInt:
+        return bit_length(self.n_nonzero_coeff - 1)
+
+    @cached_property
+    def alternates_bitsize(self) -> SymbolicInt:
+        return total_bits(self.selection_registers)
+
+    @cached_property
+    def keep_bitsize(self) -> SymbolicInt:
+        return self.mu
+
+    @cached_property
+    def selection_bitsize(self) -> SymbolicInt:
+        return total_bits(self.selection_registers)
+
+    @cached_property
+    def junk_registers(self) -> Tuple[Register, ...]:
+        return tuple(
+            Signature.build(
+                sigma_mu=self.sigma_mu_bitsize,
+                sparse_index=self.sparse_index_bitsize,
+                alt=self.alternates_bitsize,
+                keep=self.keep_bitsize,
+                less_than_equal=1,
+            )
+        )
+
+    @cached_property
+    def qrom_bloq(self) -> QROM:
+        return QROM(
+            (self.index, self.alt, self.keep),
+            (self.sparse_index_bitsize,),
+            (self.selection_bitsize, self.alternates_bitsize, self.keep_bitsize),
+        )
+
+    def build_composite_bloq(self, bb: 'BloqBuilder', **soqs: 'SoquetT') -> Dict[str, 'SoquetT']:
+        soqs['sparse_index'] = bb.add(
+            PrepareUniformSuperposition(self.n_nonzero_coeff), target=soqs['sparse_index']
+        )
+        if self.mu == 0:
+            soqs['sparse_index'], soqs['selection'] = bb.add(
+                QROM((self.index,), (self.sparse_index_bitsize,), (self.selection_bitsize,)),
+                selection=soqs['sparse_index'],
+                target0_=soqs['selection'],
+            )
+        else:
+            soqs['sigma_mu'] = bb.add(OnEach(self.sigma_mu_bitsize, Hadamard()), q=soqs['sigma_mu'])
+            soqs['sparse_index'], soqs['selection'], soqs['alt'], soqs['keep'] = bb.add_t(
+                self.qrom_bloq,
+                selection=soqs['sparse_index'],
+                target0_=soqs['selection'],
+                target1_=soqs['alt'],
+                target2_=soqs['keep'],
+            )
+            soqs['keep'], soqs['sigma_mu'], soqs['less_than_equal'] = bb.add_t(
+                LessThanEqual(self.mu, self.mu),
+                x=soqs['keep'],
+                y=soqs['sigma_mu'],
+                target=soqs['less_than_equal'],
+            )
+            soqs['less_than_equal'], soqs['alt'], soqs['selection'] = bb.add_t(
```

Comment:
Shouldn't you be swapping ind and alt, not alt and selection?

Reply:
the input selection register is what I treat as `ind`. I create an additional junk register `sparse_index` which is put in superposition at the start (unlike `selection` in the other bloq).

Reply:
ohh, very confusing lol.

Reply:
I should have read the docstring